Adding shared flags for D3D11 and 12 resources.

tksgmsy opened the pull request:
This change adds 'isSharedAcrossDevice' and 'isSharedAcrossAdapter' to TextureDesc and BufferDesc.
In D3D11 and 12, applying these flags will result in applying shared resource flags.
In VK, applying this flag will produce an error because it's too different the way to share resource against D3Ds.
VK needs information about the other hand's API to apply sharing flags.

tksgmsy commented:
I have read the CLA Document and I hereby sign the CLA

apanteleev commented:
Can you please add some comments to the new fields in the `TextureDesc` and `BufferDesc` structure declarations? Just a brief description of what these fields mean and their restrictions (i.e. that they are not supported on Vulkan).

Also I don't quite understand what it means that `VkExternalMemoryHandleTypeFlagBits requires explicit definition of the other hand of the API`. Never used cross-adapter/device resource sharing.

tksgmsy commented:
Just two booleans are bit too vague to represent the meaning of the flags, so I introduced new bit mask to explicitly apply the flag for resource shareing, and, I added a detailed comments.
`enum class SharedResourceFlags : uint32_t`

> Also I don't quite understand what it means that VkExternalMemoryHandleTypeFlagBits requires explicit definition of the other hand of the API. Never used cross-adapter/device resource sharing.
I'm not an expert on resource sharing on vulkan, but as far as I checked the specs, but seems VkExportMemoryAllocateInfo is needed in pNext chain of VkMemoryAllocateInfo when allocating a memory that is used by a shared resource. So I think a combination of VkExternalMemoryHandleTypeFlagBits mask is needed to be set when allocating a resource.

apanteleev commented:
Thanks!
Merged through squash, with minor cleanup, in 12f3e12eb29b681e99a53af7347c60c0733a6ef9 
Closing
